src/connection.rs

```diff
@@ -17,484 +17,505 @@
+    mode: Mode,
+    is_dead: bool,
+    config: Config,
+    streams: BTreeMap<stream::Id, StreamEntry>,
+    resource: Fuse<Framed<T, FrameCodec>>,
+    incoming: VecDeque<stream::Id>,
+    pending: VecDeque<RawFrame>,
+    tasks: Vec<Task>,
+    next_id: u32
 }
 
-
-/// A connection which multiplexes streams to the remote endpoint.
-pub struct Connection<T> {
-    is_dead: bool,
-    mode: Mode,
-    resource: Framed<T, FrameCodec>,
-    config: Arc<Config>,
-    id_counter: u32,
-    streams: BTreeMap<stream::Id, StreamHandle>,
-    controller: Controller,
-    stream_rx: mpsc::UnboundedReceiver<(stream::Id, Item)>,
-    stream_tx: mpsc::UnboundedSender<(stream::Id, Item)>,
-    pending: Option<RawFrame>
+impl<T> fmt::Debug for Inner<T> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Connection {{ \
+                mode: {:?}, \
+                streams: {}, \
+                incoming: {}, \
+                pending: {}, \
+                next_id: {}, \
+                tasks: {} \
+            }}",
+            self.mode,
+            self.streams.len(),
+            self.incoming.len(),
+            self.pending.len(),
+            self.next_id,
+            self.tasks.len()
+        )
+    }
 }
 
-impl<T> Connection<T>
+impl<T> Inner<T>
 where
     T: AsyncRead + AsyncWrite
 {
-    /// Create a new connection either in client or server mode.
-    pub fn new(resource: T, config: Arc<Config>, mode: Mode) -> Self {
-        debug!("new connection");
-        let controller = {
-            let (tx, rx) = mpsc::channel(1024);
-            Controller {
-                sender: Ctrl::new(config.clone(), tx),
-                receiver: rx.fuse()
-            }
-        };
-        let (stream_tx, stream_rx) = mpsc::unbounded();
-        Connection {
+    fn new(resource: T, config: Config, mode: Mode) -> Self {
+        Inner {
             mode,
             is_dead: false,
-            resource: Framed::new(resource, FrameCodec::new()),
             config,
-            id_counter: match mode {
+            streams: BTreeMap::new(),
+            resource: Framed::new(resource, FrameCodec::new()).fuse(),
+            incoming: VecDeque::new(),
+            pending: VecDeque::new(),
+            tasks: Vec::new(),
+            next_id: match mode {
                 Mode::Client => 1,
                 Mode::Server => 2
-            },
-            streams: BTreeMap::new(),
-            controller,
-            stream_rx,
-            stream_tx,
-            pending: None
+            }
         }
     }
 
-    /// Get a control handle which allows to open new streams.
-    pub fn control(&self) -> Ctrl {
-        self.controller.sender.clone()
+    fn next_stream_id(&mut self) -> Result<stream::Id, ConnectionError> {
+        if self.next_id > u32::MAX - 2 {
+            return Err(ConnectionError::NoMoreStreamIds)
+        }
+        let proposed = stream::Id::new(self.next_id);
+        self.next_id += 2;
```

Comment:
I'd suggest using `checked_add(2)` here, and return `ConnectionError::NoMoreStreamIds` if it fails, and remove the `if` above.